### ISSUE TITLE
csrf token per session

### ISSUE THREAD

**Issue by Conni2461** [edward-shen/actix-csrf#9]
So i have a resource like `POST /user/{uuid}` which allows me to edit a user, so if i now submit the form everything works like a charm.

Now i open two different users in two tabs. If i now submit the form of the first tab that i have opened, i get an "422
Unprocessable Entity". I can still submit the second tab, that works perfectly fine.

Does that make sense or should i provided a minimal example?

I dont think that is correct and i am currently not sure how to mitigate this issue ^^

**Comment by Conni2461** [comment]
I skimmed though the django csrftoken middleware implementation and it seems like that django only rotates the token on login and offers a public function to rotate it manually. Couldnt we do something like this here as well? Like configure the middleware with an option to disable auto rotate (then we continue to use the token that is stored in either as cookie or in the session (feature is currently not implemented but would be something we should do)). And then we have a function that rotate the token i can manually call on login ?!

Ref: django implementation https://docs.djangoproject.com/en/3.2/_modules/django/middleware/csrf/

Edit: the blog post you linked: https://cheatsheetseries.owasp.org/cheatsheets/Cross-Site_Request_Forgery_Prevention_Cheat_Sheet.html#synchronizer-token-pattern also mentions:
> They can be generated once per user session or for each request.

So what i kinda need is the first, (generated once per user session). I could work on a solution if you are busy, but i'm kinda also busy but need a solution, sooner rather than later :laughing: everything sucks i know.

And sorry for the somewhat weird issue ;)

**Comment by edward-shen** [comment]
Hey! Thanks again for the interest :)

I'm definitely open to this feature being implemented, but with the holiday season and the current situation at my company I'm not in a position where I can freely contribute time into implementation.

I'm unfortunately not able to make any promises on when I would get time to work on this (definitely not any time within the next few months), so if you need this for something that's time sensitive (e.g. project deadlines), I would suggest that you implement the solution yourself. Of course, I would be happy to review the code changes and publish an updated crate version with just these changes.

**Comment by Conni2461** [comment]
Imagine having sensible deadlines. More like: "Can this be fixed till tomorrow? Thanks" ^^

But yeah thanks for your support. I'll look into implementing it. I am kinda happy that you are too busy to implement it because the more i read about it the more interested i became in actually solving it myself ^^

Thank you, I'll get back to you when I have something to review :)

**Comment by ghost** [comment]
The way I see it, we can do this one of three ways. My personal preference goes toward option 2, but I'm happy to hear thoughts on this.

##  1. Using `actix_session` and making it our responsibility to register it
We "just" need something to store the session, and the easiest way to do this is to simply use `actix_session`. They've already figured out how to do this securely, so we don't have to. Technically there Is one way, by making a `CsrfSession` middleware that initializes BOTH `actix_session::SessionMiddleware` and `CsrfMiddleware` in the correct order and we tell the user that they can pick and choose which session store is used on their own (based on [this](https://docs.rs/actix-session/latest/actix_session/index.html#choosing-a-backend)). The user can then still use `actix_session` for any other things they may want to store in the session, but must not initialize it themselves. (idk what side effects it could have to initialize it twice. May cause weird issues?)


## 2. Using `actix_session` and making it the user's responsibility to register it
By far the easiest solution is to assume that the user must set `actix_session` up on their own. This way we simply make the assumption "config for `per_session` is set, just assume `actix_session` is available". This would require thorough documentation on the fact that `CsrfMiddleware` must be registered AFTER `actix_session::SessionMiddleware`. My partner has a version of this idea in their current project, [here](https://codeberg.org/alexnya/purrbox/src/branch/mistress/src/csrf/mod.rs). This was basically created because they wanted this exact functionality but (understandably) didn't want to wait for contributions to get accepted.

## 3. The DIY solution
We basically make our own, purpose built, session store. This... might actually require a bit more effort and could end up far less secure than just using `actix_session`. It may also possibly interfere with a user that is already using `actix_session`, If we're using cookies to store said session.

**Comment by edward-shen** [comment]
Thanks for your thoughts! I'm aligned with you on option 2 -- with the additions of feature-gating it and panicking if the SessionMiddleware isn't initialized before CrsfMiddleware is.

A pattern I've also seen is to take some sort of proof that the SessionMiddleware has been registered, but unfortunately we don't have that available here.

**Comment by ghost** [comment]
> Thanks for your thoughts! I'm aligned with you on option 2 -- with the additions of feature-gating it and panicking if the SessionMiddleware isn't initialized before CrsfMiddleware is.
> 
> A pattern I've also seen is to take some sort of proof that the SessionMiddleware has been registered, but unfortunately we don't have that available here.

Agreed. Would you prefer a separate `CsrfSessionMiddleware` or just adding the logic into the current `CsrfMiddleware` behind a feature gate? 

For ease of use I kinda like the second one. It would also save on a lot of duplicate logic. Though for clarity and documentation purposes, the first one is far better. Personally I don't mind either implementation, they both have  their ups and downs.

**Comment by edward-shen** [comment]
Feel free to add it to the current middleware -- it's configured via Builder API already so augmenting it should be simple.
